File: discrete-optimization/02-knapsack/Solver.java
import java.io.*;
import java.util.List;
import java.util.ArrayList;
import java.util.Comparator;

class Item {
    int index;
    int value;
    int weight;
    int pick;

    public Item(int index, int value, int weight, int pick) {
        this.index = index;
        this.value = value;
        this.weight = weight;
        this.pick = pick;
    }

    public void updatePick(int newPick) {
        this.pick = newPick;
    }
}

/**
 * The class <code>Solver</code> is an implementation of a greedy algorithm to solve the knapsack problem.
 *
 */
public class Solver {

    /**
     * The main class
     */
    public static void main(String[] args) {
        try {
            solve(args);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    /**
     * Read the instance, solve it, and print the solution in the standard output
     */
    public static void solve(String[] args) throws IOException {
        String fileName = null;

        // get the temp file name
        for(String arg : args){
            if(arg.startsWith("-file=")){
                fileName = arg.substring(6);
            }
        }
        if(fileName == null)
            return;

        // read the lines out of the file
        List<String> lines = new ArrayList<String>();

        BufferedReader input =  new BufferedReader(new FileReader(fileName));
        try {
            String line = null;
            while (( line = input.readLine()) != null){
                lines.add(line);
            }
        }
        finally {
            input.close();
        }
        // parse the data in the file
        String[] firstLine = lines.get(0).split("\\s+");
        int items = Integer.parseInt(firstLine[0]);
        int capacity = Integer.parseInt(firstLine[1]);

        int[] values = new int[items];
        int[] weights = new int[items];

        for(int i=1; i < items+1; i++){
          String line = lines.get(i);
          String[] parts = line.split("\\s+");

          values[i-1] = Integer.parseInt(parts[0]);
          weights[i-1] = Integer.parseInt(parts[1]);
        }

        // a trivial greedy algorithm for filling the knapsack
        // it takes items in-order until the knapsack is full
        int value = 0;
        int weight = 0;
        int[] taken = new int[items];



        // for (int i = 1; i <= item_count; i++) {
        //     String line = lines[i];
        //     String[] parts = line.split(" ");
        //     items.add(new Item(i - 1, Integer.parseInt(parts[0]), Integer.parseInt(parts[1]), 0));
        //     max_estimate += Integer.parseInt(parts[0]);
        // }

        // items.sort(Comparator.comparingDouble(item -> (double) item.value / item.weight).reversed());
        // for(int i=0; i < items; i++){
        //     if(weight + weights[i] <= capacity){
        //         taken[i] = 1;
        //         value += values[i];
        //         weight += weights[i];
        //     } else {
        //         taken[i] = 0;
        //     }
        // }

        // prepare the solution in the specified output format
        System.out.println(value+" 0");
        for(int i=0; i < items; i++){
            System.out.print(taken[i]+" ");
        }
        System.out.println("");
    }

    private static void go(List<Integer> path, int value, int room, int estimate, List<Item> items, int[] result) {
        int max_value = 0;
        List<Object[]> stack = new ArrayList<>();
        stack.add(new Object[]{path, value, room, estimate});

        while (!stack.isEmpty()) {
            Object[] current = stack.remove(stack.size() - 1);
            List<Integer> currentPath = (List<Integer>) current[0];
            int currentValue = (int) current[1];
            int currentRoom = (int) current[2];
            int currentEstimate = (int) current[3];

            if (currentPath.size() == items.size()) {
                if (currentValue > max_value) {
                    max_value = currentValue;
                    result[0] = currentValue;
                    result[1] = currentPath.get(currentPath.size() - 1);
                }
                continue;
            }

            Item item = items.get(currentPath.size());

            int estimateAll = estimate;

            if (currentRoom - item.weight >= 0) {
                List<Integer> newPath = new ArrayList<>(currentPath);
                newPath.add(1);
                stack.add(new Object[]{newPath, currentValue + item.value, currentRoom - item.weight, estimate});
            }

            if (item.pick == 1) {
                estimateAll -= item.value;
            }

            if (estimateAll >= result[0]) {
                List<Integer> newPath = new ArrayList<>(currentPath);
                newPath.add(0);
                stack.add(new Object[]{newPath, currentValue, currentRoom, estimateAll});
            }
        }
    }

    private static int[] sortPathByItemIndex(int[] path, List<Item> items) {
        int[] sortedPath = new int[path.length];
        int index = 0;
        for (int itemIndex : path) {
            for (int i = 0; i < items.size(); i++) {
                if (items.get(i).index == itemIndex) {
                    sortedPath[index++] = itemIndex;
                    break;
                }
            }
        }
        return sortedPath;
    }

}

File: discrete-optimization/02-knapsack/solver_.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
from collections import namedtuple
Item = namedtuple("Item", ['index', 'value', 'weight', 'pick'])
from numba import njit

# class Item:
    # def __init__(self, index, value, weight, pick):
    #     self.index = index
    #     self.value = value
    #     self.weight = weight
    #     self.pick = pick

    # def update_pick(self, new_pick):
    #     self.pick = new_pick

    # def print(self):
    #     print("index", self.index, "value", self.value, "weight", self.weight, "pick", self.pick)


def sort_path_by_item_index(path, items):
    item_indices = {item.index: index for index, item in enumerate(items)}
    sorted_path = sorted(path, key=lambda x: item_indices[x])
    return sorted_path

def solve_it(input_data):

    # Modify this code to run your optimization algorithm

    # parse the input
    lines = input_data.split('\n')

    firstLine = lines[0].split()
    item_count = int(firstLine[0])
    capacity = int(firstLine[1])
    max_estimate = 0

    items = []

    for i in range(1, item_count+1):
        line = lines[i]
        parts = line.split()
        items.append(Item(i-1, int(parts[0]), int(parts[1]), 0))
        # max_estimate += int(parts[0])

    left_room = capacity

    while left_room>0:
        max_value_density = max(items, key=lambda x: x.value/x.weight if x.pick == 0 else 0)
        if max_value_density.weight <= left_room:
            left_room -=  max_value_density.weight
            # max_value_density.update_pick(1)
            max_value_density = max_value_density._replace(pick=1)
            max_estimate += max_value_density.value
        else:
            max_estimate += (max_value_density.value/max_value_density.weight) * left_room
            left_room = 0

    # 排序
    items.sort(key=lambda x: x.value / x.weight, reverse=True)

    value = 0
    taken = [0]*len(items)
    estimate_all = 0
    result = {}

    def go(path, value, room, estimate):
        nonlocal estimate_all
        nonlocal result
        max_value = 0
        stack = [(path, value, room, estimate)]

        while stack:
            path, value, room, estimate = stack.pop()

            if len(path) == len(items):
                if value > max_value:
                    max_value = value
                    # 计算 weight
                    result = {
                        "path": path,
                        "value": value,
                    }
                    estimate_all = max(value, estimate_all)
                continue

            item = items[len(path)]

            _estimate = estimate

            if room  - item.weight >= 0:
                stack.append((path + [1], value + item.value, room - item.weight, estimate))

            if item.pick == 1:
                _estimate -=  item.value

            # 这句再想想问题
            if _estimate >= estimate_all:
                # print("estimate_all", estimate_all)
                stack.append((path + [0], value, room, _estimate))

    go([], value, capacity, max_estimate)

    # max_result = max(result, key=lambda x: x['value'])
    value = result["value"]
    taken = sort_path_by_item_index(result["path"], items)

    # prepare the solution in the specified output format
    output_data = str(value) + ' ' + str(0) + '\n'
    output_data += ' '.join(map(str, taken))
    return output_data


if __name__ == '__main__':
    import sys
    if len(sys.argv) > 1:
        file_location = sys.argv[1].strip()
        with open(file_location, 'r') as input_data_file:
            input_data = input_data_file.read()
        print(solve_it(input_data))
    else:
        print('This test requires an input file.  Please select one from the data directory. (i.e. python solver.py ./data/ks_4_0)')

